### PR TITLE
fix: don't make a copy of the webui unet state dict

### LINKED ISSUE #5: Checkpoints aren't shared
Would be nice to have a node "WebuiCheckpointLoader" that can use the checkpoint that's already in memory.

### PR DESCRIPTION
**Pull request by ljleb**
- [x] unet proxy
- [x] vae proxy
- [x] clip proxy
- [x] ~~make all types of supported models work~~ (too big a task, will follow up with another PR to keep things manageable)
- [x] ~~lora~~ (will be implemented later)
- [x] cleanup

closes #5

### CODE REVIEW

File: lib_comfyui/webui_proxies.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+        return res
+
+
+def sd_model_apply(*args, **kwargs):
+    args = torch_utils.deep_to(args, shared.sd_model.device)
+    kwargs = torch_utils.deep_to(kwargs, shared.sd_model.device)
+    with devices.autocast(), torch.no_grad():
+        res = shared.sd_model.model(*args, **kwargs)
+        return res.detach().cpu().share_memory_()
```
Comment (ljleb) [inline lib_comfyui/webui_proxies.py:94]: note to self: only cpu() useful here

**Comment by ljleb** [comment]
There's a RAM overhead I can't get rid of for some reason.

File: comfyui_custom_nodes/webui_checkpoint_loader.py [hunk 2]
```diff
@@ -1,26 +1,5 @@
...
-    def __exit__(self, type, value, traceback):
-        utils.load_torch_file = self.original_loader
-
-    @staticmethod
-    def load_torch_file_patch(state_dict, safe_load=False):
-        return state_dict.copy()
-
-    def load(self):
-        return load_checkpoint_guess_config(self.state_dict)
+from lib_comfyui.webui_proxies import WebuiModelPatcher, WebuiModelProxy, WebuiClipWrapper, WebuiClipProxy, WebuiVaeWrapper, WebuiVaeProxy
```
Comment (John-WL) [inline comfyui_custom_nodes/webui_checkpoint_loader.py:1]:
```py
import (
-
-
-
-
-
)
```

File: lib_comfyui/ipc.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    function = module
+    for name in qualified_name.split('.'):
+        function = getattr(function, name)
+    return function(*args, **kwargs)
+
+
+current_process_id = 'webui'
+current_process_callback_listeners = {
+    'webui': parallel_utils.ProducerHandler(parallel_utils.SynchronizingQueue(call_fully_qualified)),
```
Comment (John-WL) [inline lib_comfyui/ipc.py:38]: rename to "CallbackListener"?

File: lib_comfyui/webui_proxies.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+            padding_tokens_count = ((tokens_count // clip_max_len) + 1) * clip_max_len
+            for token_i in range(padding_tokens_count):
+                actual_id = token_i if token_i < tokens_count else 0
+                weighted_tokens[token_i // clip_max_len][token_i % clip_max_len] += (actual_id,)
+
+        return weighted_tokens
+
+    @property
+    def layer_idx(self):
+        clip_skip = self.cond_stage_model.config.num_hidden_layers - webui_settings.opts.CLIP_stop_at_last_layers
```
Comment (John-WL) [inline lib_comfyui/webui_proxies.py:222]: are you sure this is fine?

**Comment by John-WL** [comment]
what the heck, nice job

File: lib_comfyui/webui_proxies.py (hunk 4 above)
Comment (ljleb) [reply lib_comfyui/webui_proxies.py]: thanks, now it should be implemented with respect to the clip skip node in comfy

**Comment by ljleb** [comment]
I replaced all but one hard errors with soft errors in the console so that people don't have to keep switching workflows to get the webui model to work. The only hard check made for now is the model type, to prevent unsupported models from crashing with incomprehensible errors.

**Review by John-WL — APPROVED** [review]

**Pull request merged**